dashboard/app/asset_storage.go

```diff
@@ -0,0 +1,368 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/syzkaller/dashboard/dashapi"
+	"github.com/google/syzkaller/pkg/asset"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/v2"
+	db "google.golang.org/appengine/v2/datastore"
+	"google.golang.org/appengine/v2/log"
+)
+
+// TODO: decide if we really want to save job-related assets right now.
+//
+// The problem is that we cannot first report the job status and then upload the assets
+// because an email will already be sent by that time.
+//
+// With the existing asset types it does not seem to be worth it to solve those problems.
+// Probably once we have things like code dumps it will start to make some sense.
+
+func appendBuildAssets(c context.Context, ns string, buildID string, assets []Asset) (*Build, error) {
+	var retBuild *Build
+	tx := func(c context.Context) error {
+		build, err := loadBuild(c, ns, buildID)
+		if err != nil {
+			return err
+		}
+		retBuild = build
+		appendedOk := false
+		var appendErr error
+		for _, newAsset := range assets {
+			appendErr = build.AppendAsset(newAsset)
+			if appendErr == nil {
+				appendedOk = true
+			}
+		}
+		// It took quite a number of resources to upload the files, so we return success
+		// even if we managed to save at least one of the new assets.
+		if !appendedOk {
+			return fmt.Errorf("failed to append all assets, last error %w", appendErr)
+		}
+		if _, err := db.Put(c, buildKey(c, ns, buildID), build); err != nil {
+			return fmt.Errorf("failed to put build: %w", err)
+		}
+		log.Infof(c, "updated build: %#v", build)
+		return nil
+	}
+	if err := db.RunInTransaction(c, tx, &db.TransactionOptions{}); err != nil {
+		log.Errorf(c, "failed to update build: %v", err)
+		return nil, err
+	}
+	return retBuild, nil
+}
+
+var ErrAssetDuplicated = errors.New("an asset of this type is already present")
+
+func (build *Build) AppendAsset(addAsset Asset) error {
+	typeInfo := asset.GetTypeDescription(addAsset.Type)
+	if typeInfo == nil {
+		return fmt.Errorf("unknown asset type")
+	}
+	if !typeInfo.AllowMultiple {
+		for _, obj := range build.Assets {
+			if obj.Type == addAsset.Type {
+				return ErrAssetDuplicated
+			}
+		}
+	}
+	build.Assets = append(build.Assets, addAsset)
+	return nil
+}
+
+func parallelize(errorStart string, funcs ...func() error) error {
```

Comment:
`errgroup.Group` may be an existing altearntive:
https://www.storj.io/blog/production-concurrency
https://pkg.go.dev/golang.org/x/sync/errgroup

Reply:
Indeed! Fixed.